src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+            }
+        }
+
+        if (isCompoundExpression && canEvaluate)
+        {
+            var assignmentExpression = syntaxNode as AssignmentExpressionSyntax;
+            var leftHandSideIdentifier = assignmentExpression.Left.TrimParenthesis() as IdentifierNameSyntax;
+            canEvaluate &= variableValues.TryGetValueSafe(leftHandSideIdentifier.Identifier.ValueText, out var leftVariableValue);
+
+            if (canEvaluate)
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                syntaxNodeToVariableValueMap.Add(leftHandSideIdentifier, leftVariableValue);
+            }
+        }
+
+        return new EvaluationResult(canEvaluate, syntaxNodeToVariableValueMap);
+    }
+
+    private static ExpressionAnalysisContext GenerateNewExpressionAnalysisContext(
+        SyntaxNode syntaxNode,
+        Dictionary<SyntaxNode, VariableValue> builderNodeToVariableValue)
+    {
+        var syntaxKind = syntaxNode.Kind();
+        var argumentTypeName = "";
+        var constantMappers = new List<ConstantsMapper>();
+        var nodesRemapping = new Dictionary<SyntaxNode, SyntaxNode>();
+
+        foreach (var builderNode in builderNodeToVariableValue.Keys)
+        {
+            var variableValue = builderNodeToVariableValue[builderNode];
+            var analysisContext = variableValue.ExpressionAnalysisContext;
+            var rewrittenBuildersNode = analysisContext.Node.RewrittenExpression;
+            var constantsMapper = analysisContext.Node.ConstantsRemapper;
+            argumentTypeName = analysisContext.Node.ArgumentTypeName;
+            nodesRemapping.Add(builderNode, rewrittenBuildersNode);
+            constantMappers.Add(constantsMapper);
+        }
+
+        var rewrittenSyntaxNode = syntaxNode.ReplaceNodes(nodesRemapping.Keys, (n, _) => nodesRemapping[n]);
+
+        if (IsCompoundExpression(syntaxKind))
+        {
+            var rewrittenAssignmentExpression = rewrittenSyntaxNode as AssignmentExpressionSyntax;
+            var binaryExpressionSyntaxKind = GenerateBinaryExpressionSyntaxKind(syntaxKind);
+            rewrittenSyntaxNode = SyntaxFactory.BinaryExpression(binaryExpressionSyntaxKind, rewrittenAssignmentExpression.Left, rewrittenAssignmentExpression.Right);
+        }
+
+        var rewrittenBuildersExpression = rewrittenSyntaxNode switch
+        {
+            AssignmentExpressionSyntax @assignmentExpression => @assignmentExpression.Right,
+            VariableDeclaratorSyntax @variableDeclarator => @variableDeclarator.Initializer.Value,
+            _ => rewrittenSyntaxNode
+        };
+
+        var (finalBuildersExpression, finalConstantMapper) = ConstantsMapperComposer.Compose(rewrittenBuildersExpression, constantMappers);
+        return new ExpressionAnalysisContext(new ExpressionAnalysisNode(rewrittenBuildersExpression, argumentTypeName, finalBuildersExpression, finalConstantMapper, new Location[] { }));
```

Comment:
No need to create new Location[] { } instance if it's empty.
Also don't  we need location here?

Reply:
We don't need location. Given an expression z = a & b & c, We compute the new ExpressionAnalysisContext for "z" but since "z" is on LHS, we don't actually "display" the diagnostic showing MQL. Only in future declarations/assignments, when "z" appears on the RHS, then we would add to this location array and display the MQL in those places.